lib/client-request.js

```diff
@@ -0,0 +1,153 @@
+const { createConnection } = require('bare-tcp')
+const HTTPOutgoingMessage = require('./outgoing-message')
+const HTTPIncomingMessage = require('./incoming-message')
+const errors = require('./errors')
+
+module.exports = class HTTPClientRequest extends HTTPOutgoingMessage {
+  constructor (opts = {}, onresponse = null) {
+    const socket = createConnection(opts)
+
+    super(socket)
+
+    if (typeof opts === 'function') {
+      onresponse = opts
+      opts = {}
+    }
+
+    this.method = opts.method || 'GET'
+    this.path = opts.path || '/'
+    const host = opts.host || 'localhost'
+    const port = opts.port || 80
+
+    this.headers = { host: host + ':' + port }
+
+    this._socket = socket
+    this._replied = false
+
+    this._pendingFinal = null
+
+    if (onresponse) this.once('response', onresponse)
+
+    socket
+      .on('close', this._onclose.bind(this))
+      .on('data', this._ondata.bind(this))
+      .on('end', this._onend.bind(this))
+      .on('error', this._onerror.bind(this))
+  }
+
+  _header () {
+    let h = `${this.method} ${this.path} HTTP/1.1\r\n`
+
+    for (const name of Object.keys(this.headers)) {
+      const n = name.toLowerCase()
+      const v = this.headers[name]
+
+      h += `${httpCase(n)}: ${v}\r\n`
+    }
+
+    h += '\r\n'
+
+    return h
+  }
+
+  _final (cb) {
+    if (this.headersSent === false) this.flushHeaders()
+
+    this._pendingFinal = cb
+  }
+
+  _ondata (data) {
+    const blankLine = Buffer.from([13, 10, 13, 10])
+    const i = data.indexOf(blankLine)
+
+    if (i === -1) return this._socket.destroy()
+
+    const head = data.subarray(0, i)
+    const messageBody = data.subarray(i + blankLine.byteLength)
+
+    this._onresponse(head, messageBody)
+  }
```

Comment:
This assumes that a single TCP message corresponds to a full HTTP header, which isn't necessarily the case. TCP doesn't frame messages and is free to split writes into as many or few messages as it finds appropriate.

You can see an example of how to correctly stitch together the TCP messages in the `ServerConnection` class, whose `_ondata()` handler uses a persistent buffer (`_buffer`) to gradually build up the complete HTTP header and only then forwards to the  `_onrequest()` handler.